enzyme/Enzyme/GradientUtils.cpp

```diff
@@ -2878,29 +2887,48 @@ Value *GradientUtils::invertPointerM(Value *const oval, IRBuilder<> &BuilderM,
       Value *val = invertPointerM(CD->getOperand(i), BuilderM);
       Vals.push_back(cast<Constant>(val));
     }
-    return ConstantArray::get(CD->getType(), Vals);
+
+    auto rule = [&CD](ArrayRef<Constant *> Vals) {
+      return ConstantArray::get(CD->getType(), Vals);
+    };
+
+    return applyChainRule(CD->getType(), Vals, BuilderM, rule);
   } else if (auto CD = dyn_cast<ConstantStruct>(oval)) {
     SmallVector<Constant *, 1> Vals;
     for (size_t i = 0, len = CD->getNumOperands(); i < len; i++) {
       Vals.push_back(
           cast<Constant>(invertPointerM(CD->getOperand(i), BuilderM)));
     }
-    return ConstantStruct::get(CD->getType(), Vals);
+
+    auto rule = [&CD](ArrayRef<Constant *> Vals) {
+      return ConstantStruct::get(CD->getType(), Vals);
+    };
+
+    return applyChainRule(CD->getType(), Vals, BuilderM, rule);
   } else if (auto CD = dyn_cast<ConstantVector>(oval)) {
     SmallVector<Constant *, 1> Vals;
     for (size_t i = 0, len = CD->getNumOperands(); i < len; i++) {
       Vals.push_back(
           cast<Constant>(invertPointerM(CD->getOperand(i), BuilderM)));
     }
-    return ConstantVector::get(Vals);
+
+    auto rule = [](ArrayRef<Constant *> Vals) {
+      return ConstantVector::get(Vals);
+    };
+
+    return applyChainRule(CD->getType(), Vals, BuilderM, rule);
   } else if (isa<ConstantData>(oval) && nullShadow) {
-    return Constant::getNullValue(oval->getType());
+    auto rule = [&oval]() { return Constant::getNullValue(oval->getType()); };
+
+    return applyChainRule(oval->getType(), BuilderM, rule);
   }
 
   if (isConstantValue(oval)) {
     // NOTE, this is legal and the correct resolution, however, our activity
     // analysis honeypot no longer exists
-    return getNewFromOriginal(oval);
+    auto rule = [&]() { return getNewFromOriginal(oval); };
```

Comment:
Might as well do the newFromOriginal query once, outside the chain rule

Reply:
fixed